src/tiledbsoma_ml/io_batch_iterable.py

```diff
@@ -0,0 +1,140 @@
+# Copyright (c) 2021-2024 The Chan Zuckerberg Initiative Foundation
+# Copyright (c) 2021-2024 TileDB, Inc.
+#
+# Licensed under the MIT License.
+
+import gc
+import logging
+import time
+from typing import Iterable, Iterator, Optional, Sequence, Tuple
+
+import attrs
+import numpy as np
+import pandas as pd
+import pyarrow as pa
+from tiledbsoma import DataFrame, IntIndexer, SparseNDArray
+
+from tiledbsoma_ml._csr import CSR_IO_Buffer
+from tiledbsoma_ml._utils import EagerIterator, batched
+from tiledbsoma_ml.common import NDArrayJoinId
+from tiledbsoma_ml.query_ids import Chunks
+
+logger = logging.getLogger("tiledbsoma_ml.io_batch_iterable")
+IOBatch = Tuple[CSR_IO_Buffer, pd.DataFrame]
+"""Tuple type emitted by |IOBatchIterable|, containing ``X`` rows (as a |CSR_IO_Buffer|) and ``obs`` rows
+(|pd.DataFrame|)."""
+
+
+@attrs.define(frozen=True)
+class IOBatchIterable(Iterable[IOBatch]):
+    """Given a list of ``obs_joinid`` |Chunks|, re-chunk them into (optionally shuffled) |IOBatch|'s".
+
+    An |IOBatch| is a tuple consisting of a batch of rows from the ``X`` |SparseNDArray|, as well as the corresponding
+    rows from the ``obs`` |DataFrame|. The ``X`` rows are returned in an optimized |CSR_IO_Buffer|.
+    """
+
+    chunks: Chunks
+    io_batch_size: int
+    obs: DataFrame
+    var_joinids: NDArrayJoinId
+    X: SparseNDArray
+    obs_column_names: Sequence[str] = ("soma_joinid",)
+    seed: Optional[int] = None
+    shuffle: bool = True
+    use_eager_fetch: bool = True
+
+    @property
+    def io_batch_ids(self) -> Iterable[Tuple[int, ...]]:
+        """Re-chunk ``obs_joinids`` according to the desired ``io_batch_size``."""
+        return batched(
+            (joinid for chunk in self.chunks for joinid in chunk),
+            self.io_batch_size,
+        )
+
+    def __iter__(self) -> Iterator[IOBatch]:
+        """Emit |IOBatch|'s."""
+        # Because obs/var IDs have been partitioned/split/shuffled upstream of this class, this RNG does not need to be
+        # identical across sub-processes, but seeding is supported anyway, for testing/reproducibility.
+        shuffle_rng = np.random.default_rng(self.seed)
+        X = self.X
+        context = X.context
+        obs_column_names = (
+            list(self.obs_column_names)
+            if "soma_joinid" in self.obs_column_names
+            else ["soma_joinid", *self.obs_column_names]
+        )
+        # Round-trip though tuple avoids `TypeError: IntIndexer only supports array of type int64`.
+        # TODO: debug / work around that error; serde'ing the ndarray apparently results in a second np.int64 instance, that fails reference equality check vs. the version from the worker-process.
+        var_joinids = np.array(tuple(self.var_joinids))
```

Comment:
See my earlier comment on this - it both is a fully memory copy, and should not be necessary. Would be good to root-cause the issue here (either in this PR, or in a follow-on), as one of the critical characteristics of this loader is keeping memory use "low", thereby enabling more parallel workers